Examples/Algorithms/TrackFinding/src/HoughTransformSeeder.cpp

```diff
@@ -0,0 +1,529 @@
+    // groupByModule only works with geometry containers, not with an
+    // arbitrary range. do the equivalent grouping manually
+    auto groupedByModule = makeGroupBy(range, detail::GeometryIdGetter());
+
+    for (auto [moduleGeoId, moduleSourceLinks] : groupedByModule) {
+      // find corresponding surface
+      const Acts::Surface* surface =
+          m_cfg.trackingGeometry->findSurface(moduleGeoId);
+      if (not surface) {
+        ACTS_ERROR("Could not find surface " << moduleGeoId);
+        return ProcessCode::ABORT;
+      }
+
+      for (auto& sourceLink : moduleSourceLinks) {
+        // extract a local position/covariance independent from the concrecte
+        // measurement content. since we do not know if and where the local
+        // parameters are contained in the measurement parameters vector, they
+        // are transformed to the bound space where we do know their location.
+        // if the local parameters are not measured, this results in a
+        // zero location, which is a reasonable default fall-back.
+        auto [localPos, localCov] = std::visit(
+            [](const auto& meas) {
+              auto expander = meas.expander();
+              Acts::BoundVector par = expander * meas.parameters();
+              Acts::BoundSymMatrix cov =
+                  expander * meas.covariance() * expander.transpose();
+              // extract local position
+              Acts::Vector2 lpar(par[Acts::eBoundLoc0], par[Acts::eBoundLoc1]);
+              // extract local position covariance.
+              Acts::SymMatrix2 lcov =
+                  cov.block<2, 2>(Acts::eBoundLoc0, Acts::eBoundLoc0);
+              return std::make_pair(lpar, lcov);
+            },
+            measurements[sourceLink.get().index()]);
+
+        // transform local position to global coordinates
+        Acts::Vector3 globalFakeMom(1, 1, 1);
+        Acts::Vector3 globalPos =
+            surface->localToGlobal(ctx.geoContext, localPos, globalFakeMom);
+        double r = sqrt(globalPos[Acts::ePos0] * globalPos[Acts::ePos0] +
+                        globalPos[Acts::ePos1] * globalPos[Acts::ePos1]);
+        double phi = atan2(globalPos[Acts::ePos1], globalPos[Acts::ePos0]);
+        double z = globalPos[Acts::ePos2];
+        unsigned hitlayer = (*(m_cfg.findLayerIDMeasurement))(r);
+        std::shared_ptr<const MeasurementStruct> meas =
+            std::shared_ptr<const MeasurementStruct>(new MeasurementStruct(
+                hitlayer, phi, r, z, sourceLink.get().index()));
+        measurementStructs.push_back(meas);
+      }
+    }
+  }
+
+  static thread_local ProtoTrackContainer protoTracks;
+  protoTracks.clear();
+
+  for (auto subregion : m_cfg.m_subRegions) {
+    ActsExamples::Image m_image =
+        createImage(spacePointPtrs, measurementStructs, subregion);
+
+    for (unsigned y = 0; y < m_cfg.m_imageSize_y; y++)
+      for (unsigned x = 0; x < m_cfg.m_imageSize_x; x++)
```

Comment:
here you're looping over the whole 7000 x 216 bins, you can avoid that by looping over the MeasurementStructs (and SpacePoints, if you decide to keep them) only, which should be 1 or 2 orders of magnitude less. From each measurementStruct, you can jump into the Image. In the Image it would be enough to store a single bit for each layer that has been hit.